File: src/main/java/me/miki/shindo/utils/network/HttpUtils.java
package me.miki.shindo.utils.network;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import lombok.var;
import me.miki.shindo.logger.ShindoLogger;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Objects;

public class HttpUtils {

    private static final String ACCEPTED_RESPONSE = "application/json";
    private static final Gson gson = new Gson();

    public static JsonObject readJson(HttpURLConnection connection) {
        return gson.fromJson(readResponse(connection), JsonObject.class);
    }

    public static JsonObject postJson(String url, Object request) {
        return postJson(url, request, null);
    }

    public static JsonObject postJson(String url, Object request, Map<String, String> headers) {

        HttpURLConnection connection = setupConnection(url, UserAgents.MOZILLA, 5000, false);

        if (connection == null) {
            ShindoLogger.error("Failed to setup connection for post json");
            return null;
        }

        connection.setDoOutput(true);
        connection.addRequestProperty("Content-Type", ACCEPTED_RESPONSE);
        connection.addRequestProperty("Accept", ACCEPTED_RESPONSE);

        if (headers != null && !headers.isEmpty()) {
            for (Map.Entry<String, String> entry : headers.entrySet()) {
                if (entry.getKey() != null && entry.getValue() != null) {
                    connection.addRequestProperty(entry.getKey(), entry.getValue());
                }
            }
        }

        try {
            connection.setRequestMethod("POST");
            try (var output = connection.getOutputStream()) {
                output.write(gson.toJson(request).getBytes(StandardCharsets.UTF_8));
            }
        } catch (IOException e) {
            ShindoLogger.error("Failed to post json", e);
            return null;
        }

        return readJson(connection);
    }

    public static String readResponse(HttpURLConnection connection) {

        String redirection = connection.getHeaderField("Location");

        if (redirection != null) {
            return readResponse(Objects.requireNonNull(setupConnection(redirection, UserAgents.MOZILLA, 5000, false)));
        }

        StringBuilder response = new StringBuilder();

        try (BufferedReader br = new BufferedReader(new InputStreamReader(connection.getResponseCode() >= 400 ? connection.getErrorStream() : connection.getInputStream()))) {
            String line;
            while ((line = br.readLine()) != null) {
                response.append(line).append('\n');
            }
        } catch (IOException e) {
            ShindoLogger.error("Failed to read response", e);
        }

        return response.toString();
    }

    public static JsonObject readJson(String url, Map<String, String> headers, String userAgents) {

        try {
            HttpURLConnection connection = setupConnection(url, userAgents, 5000, false);

            if (connection == null) {
                ShindoLogger.error("Failed to setup connection for read json");
                return null;
            }

            if (headers != null && !headers.isEmpty()) {

                for (String header : headers.keySet()) {
                    connection.addRequestProperty(header, headers.get(header));
                }
            }

            InputStream is = connection.getResponseCode() != 200 ? connection.getErrorStream() : connection.getInputStream();
            BufferedReader rd = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8));

            return gson.fromJson(readResponse(rd), JsonObject.class);
        } catch (IOException e) {
            ShindoLogger.error("Failed to read json", e);
        }

        return null;
    }

    public static JsonObject readJson(String url, Map<String, String> headers) {
        return readJson(url, headers, UserAgents.MOZILLA);
    }

    private static String readResponse(BufferedReader br) {

        try {
            StringBuilder sb = new StringBuilder();
            String line;

            while ((line = br.readLine()) != null) {
                sb.append(line);
            }

            return sb.toString();
        } catch (IOException e) {
            ShindoLogger.error("Failed to read response", e);
        }

        return null;
    }


    public static boolean downloadFile(String url, File outputFile, String userAgent, int timeout, boolean useCaches) {

        url = url.replace(" ", "%20");

        try (FileOutputStream fileOut = new FileOutputStream(outputFile); BufferedInputStream in = new BufferedInputStream(setupConnection(url, userAgent, timeout, useCaches).getInputStream())) {
            org.apache.commons.io.IOUtils.copy(in, fileOut);
        } catch (Exception e) {
            ShindoLogger.error("Failed to download file", e);
            return false;
        }

        return true;
    }

    public static boolean downloadFile(String url, File outputFile, String userAgents) {
        return downloadFile(url, outputFile, userAgents, 5000, false);
    }

    public static void downloadFile(String url, File outputFile) {
        downloadFile(url, outputFile, UserAgents.MOZILLA, 5000, false);
    }


    public static HttpURLConnection setupConnection(String url, String userAgent, int timeout, boolean useCaches) {
        try {
            url = PunycodeUtils.punycode(url);
            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();

            connection.setRequestMethod("GET");
            connection.setUseCaches(useCaches);
            connection.addRequestProperty("User-Agent", userAgent);
            connection.setRequestProperty("Accept-Language", "en-US");
            connection.setRequestProperty("Accept-Charset", "UTF-8");
            connection.setReadTimeout(timeout);
            connection.setConnectTimeout(timeout);
            connection.setDoOutput(true);

            return connection;
        } catch (Exception e) {
            ShindoLogger.error("Failed to setup connection");
        }

        return null;
    }

    public static String encodeURL(String url) {
        try {
            return URLEncoder.encode(url, StandardCharsets.UTF_8.toString());
        } catch (UnsupportedEncodingException e) {
            return url;
        }
    }

    public static String decodeURL(String url) {
        try {
            return URLDecoder.decode(url, StandardCharsets.UTF_8.toString());
        } catch (UnsupportedEncodingException e) {
            return url;
        }
    }

}

File: src/main/kotlin/me/miki/shindo/utils/network/HttpUtils.kt
package me.miki.shindo.utils.network

import com.google.gson.Gson
import com.google.gson.JsonObject
import me.miki.shindo.logger.ShindoLogger
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okio.sink
import java.io.File
import java.util.concurrent.TimeUnit

object HttpUtils {
    private val JSON_TYPE = "application/json; charset=utf-8".toMediaTypeOrNull()

    private val gson = Gson()

    // Shared client — connection pooling, keep-alive, and redirect following are automatic
    private val client =
        OkHttpClient
            .Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .readTimeout(5, TimeUnit.SECONDS)
            .writeTimeout(5, TimeUnit.SECONDS)
            .followRedirects(true)
            .followSslRedirects(true)
            .build()

    // Variant with custom timeout for callers that need it (e.g. large downloads)
    private fun clientWithTimeout(timeoutMs: Long): OkHttpClient =
        client
            .newBuilder()
            .connectTimeout(timeoutMs, TimeUnit.MILLISECONDS)
            .readTimeout(timeoutMs, TimeUnit.MILLISECONDS)
            .build()

    @JvmStatic
    fun postJson(
        url: String,
        request: Any,
    ): JsonObject? = postJson(url, request, null)

    @JvmStatic
    fun postJson(
        url: String,
        request: Any,
        headers: Map<String, String>?,
    ): JsonObject? {
        val body = gson.toJson(request).toRequestBody(JSON_TYPE)

        val req =
            Request
                .Builder()
                .url(url)
                .post(body)
                .header("User-Agent", UserAgents.MOZILLA)
                .header("Accept", "application/json")
                .apply { headers?.forEach { (k, v) -> addHeader(k, v) } }
                .build()

        return executeForJson(req)
    }

    @JvmStatic
    fun readJson(
        url: String,
        headers: Map<String, String>?,
    ): JsonObject? = readJson(url, headers, UserAgents.MOZILLA)

    @JvmStatic
    fun readJson(
        url: String,
        headers: Map<String, String>?,
        userAgent: String,
    ): JsonObject? {
        val req =
            Request
                .Builder()
                .url(url)
                .get()
                .header("User-Agent", userAgent)
                .header("Accept-Language", "en-US")
                .header("Accept-Charset", "UTF-8")
                .apply { headers?.forEach { (k, v) -> addHeader(k, v) } }
                .build()

        return executeForJson(req)
    }

    @JvmStatic
    fun downloadFile(
        url: String,
        outputFile: File,
    ): Boolean = downloadFile(url, outputFile, UserAgents.MOZILLA, 5000, false)

    @JvmStatic
    fun downloadFile(
        url: String,
        outputFile: File,
        userAgent: String,
    ): Boolean = downloadFile(url, outputFile, userAgent, 5000, false)

    @JvmStatic
    fun downloadFile(
        url: String,
        outputFile: File,
        userAgent: String,
        timeoutMs: Int,
        @Suppress("UNUSED_PARAMETER") useCaches: Boolean, // OkHttp handles caching via Cache; flag kept for API compat
    ): Boolean {
        val req =
            Request
                .Builder()
                .url(url)
                .get()
                .header("User-Agent", userAgent)
                .build()

        return try {
            clientWithTimeout(timeoutMs.toLong()).newCall(req).execute().use { response ->
                if (!response.isSuccessful) {
                    ShindoLogger.error("Download failed — HTTP ${response.code} for $url")
                    return false
                }
                // Stream directly to disk without buffering the entire body in memory
                val body = response.body ?: return false
                outputFile.sink().use { fileSink ->
                    body.source().readAll(fileSink)
                }
                true
            }
        } catch (e: Exception) {
            ShindoLogger.error("Failed to download file", e)
            false
        }
    }

    @JvmStatic
    fun encodeURL(url: String): String = java.net.URLEncoder.encode(url, "UTF-8")

    @JvmStatic
    fun decodeURL(url: String): String = java.net.URLDecoder.decode(url, "UTF-8")

    // --- Private helpers ---

    private fun executeForJson(request: Request): JsonObject? {
        return try {
            client.newCall(request).execute().use { response ->
                val body = response.body?.string()
                if (body.isNullOrBlank()) {
                    if (!response.isSuccessful) {
                        ShindoLogger.error("HTTP ${response.code} with empty body for ${request.url}")
                    }
                    return null
                }
                gson.fromJson(body, JsonObject::class.java)
            }
        } catch (e: Exception) {
            ShindoLogger.error("HTTP request failed for ${request.url}", e)
            null
        }
    }
}
